### PR TITLE
Fixing zombie voice client

### LINKED ISSUE #3994: voice_client getting in bad state with non-normal ConnectionClosed codes
**Note**: I added a TLDR note at the bottom which kindof sums up my findings. If you don't have time to look at all the info I collected please at least take a look at that.

Versions
python 3.6.8
discord.py 1.3.3
Ubuntu 18.04.03 LTS

I've had some issues in the past with my bot getting into a bad state for voice channels and not being able to get out. Recently I've decided to dig in and try to figure out what's going on. The big problem with this is that it is a very inconsistent problem, and I can only debug it when the voice channel gets into a bad state, and I haven't figured out a way to trigger that manually. What I've done is I added a bunch of extra logging to voice_client.py and waited till I see the issue, and then try a couple things and see what happens. After taking a look at the logs, I think I've found a bug with voice_client.py. I'll start by explaining the issue I've been having and then show the logs that show the bug I've found.

**Note:** My apologies if I've put too much log info and text in here. I just figure for debugging these kinds of things, more information is always helpful

**Note:** As I said, I made some changes to voice_client.py. The main changes were generating a random number and saving it to voice_client.vcid when a new one is created, generating a random number at the beginning of poll_voice_ws. Those were to help keep track of which voice client and loop we're currently in. I also added some functionality that I thought might fix the problem, where if we get a ConnectionClosed that isnt in the acceptable close codes list, we try to do a disconnect before reconnecting. That's what the "Doin a disconnetaroni" is talking about. Note that that fix didn't seem to have an effect on the behavior im seeing. I've attached these voice_client.py changes here as a diff file:
[voice_client_logging.diff.txt](https://github.com/Rapptz/discord.py/files/4465529/voice_client_logging.diff.txt)

**Another Note:** This issue is not specific to just one voice channel, it sometimes happens to just one, and sometimes to a bunch, and there doesn't seem to be anything special about the voice channels its happening to.

# The Problem

The main problem that I'm having that started this investigation, is the voice client gets in a bad state where it thinks it is connected but really isnt. What is happening in voice_client.py is that in poll_voice_wc, we get a ConnectionClosed with a non-normal code (usually 1001 or 1006, but I've seen other codes in the past). What happens is that poll_voice_wc will infinitely loop trying to reconnect, but always getting the same bad code. I know that it isn't a problem with the thing its trying to connect to, because restarting the bot fixes the problem and it is able to connect just fine. For clarity and information, here's what the log looks like when this is happening:
<details>
  <summary>First it encounters the Error, and schedules a reconnect in x seconds:</summary>

```
INFO:discord.voice_client:Bad code 1006 in poll_voice_ws (channel=175384431383543809) (reconnect=True), (id=0.5783516900838339)
INFO:discord.voice_client:in poll_voice_ws loop for channel 175384431383543809 (id=0.5783516900838339) (loopid=0.16362357981923203)
ERROR:discord.voice_client:Disconnected from voice channel 175384431383543809... Reconnecting in 257.33s.
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/websockets/protocol.py", line 528, in transfer_data
    msg = yield from self.read_message()
  File "/usr/local/lib/python3.6/dist-packages/websockets/protocol.py", line 580, in read_message
    frame = yield from self.read_data_frame(max_size=self.max_size)
  File "/usr/local/lib/python3.6/dist-packages/websockets/protocol.py", line 645, in read_data_frame
    frame = yield from self.read_frame(max_size)
  File "/usr/local/lib/python3.6/dist-packages/websockets/protocol.py", line 710, in read_frame
    extensions=self.extensions,
  File "/usr/local/lib/python3.6/dist-packages/websockets/framing.py", line 100, in read
    data = yield from reader(2)
  File "/usr/lib/python3.6/asyncio/streams.py", line 672, in readexactly
    raise IncompleteReadError(incomplete, n)
asyncio.streams.IncompleteReadError: 0 bytes read on a total of 2 expected bytes

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/discord/gateway.py", line 747, in poll_event
    msg = await asyncio.wait_for(self.recv(), timeout=30.0)
  File "/usr/lib/python3.6/asyncio/tasks.py", line 358, in wait_for
    return fut.result()
  File "/usr/local/lib/python3.6/dist-packages/websockets/protocol.py", line 350, in recv
    yield from self.ensure_open()
  File "/usr/local/lib/python3.6/dist-packages/websockets/protocol.py", line 501, in ensure_open
    self.close_code, self.close_reason) from self.transfer_data_exc
websockets.exceptions.ConnectionClosed: WebSocket connection is closed: code = 1006 (connection closed abnormally [internal]), no reason

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/discord/voice_client.py", line 253, in poll_voice_ws
    await self.ws.poll_event()
  File "/usr/local/lib/python3.6/dist-packages/discord/gateway.py", line 750, in poll_event
    raise ConnectionClosed(exc, shard_id=None) from exc
discord.errors.ConnectionClosed: WebSocket connection is closed: code = 1006 (connection closed abnormally [internal]), no reason
```
</details>
<details>
  <summary>Then after x seconds, it finishes the wait, and tries to do a reconnect</summary>

```
INFO:discord.voice_client:The voice handshake is being terminated for Channel ID 175384431383543809 (Guild ID 175384431383543808)
INFO:discord.voice_client:Doin a disconnectaroni for (channel=175384431383543809) (loopid=0.16362357981923203)
INFO:discord.voice_client:Calling disconnect for channel=175384431383543809 (id=0.5783516900838339)
INFO:discord.voice_client:The voice handshake is being terminated for Channel ID 175384431383543809 (Guild ID 175384431383543808)
INFO:discord.voice_client:The voice client has been removed for Channel ID 175384431383543809 (Guild ID 175384431383543808)
INFO:discord.voice_client:Connecting to voice channel 175384431383543809...
INFO:discord.voice_client:Starting voice handshake...
```
</details>
<details>
  <summary>Then the reconnect times out (after a while)</summary>

```
INFO:discord.voice_client:The voice handshake is being terminated for Channel ID 175384431383543809 (Guild ID 175384431383543808)
INFO:discord.voice_client:The voice client has been removed for Channel ID 175384431383543809 (Guild ID 175384431383543808)
WARNING:discord.voice_client:Could not connect to voice... Retrying...
```
</details>

And then it goes back to step one with the poll_voice_ws and gets the same bad code again, and it just keeps looping with those 3 steps. One thing I noticed about this is that it seems weird that we would try to do the `await self.ws.poll_event()` even though we just timed out on the connect. it seems like that sets it up for failure. 

# The Bug/More Info

So what I noticed while messing around with this is that there's nothing to tell this poll_voice_ws loop if someone calls disconnect(). It seems the happy path is that someone calls that, and the next time `await self.ws.poll_event()` is called, it gets a normal connectionclosed code, and successfully disconnects, but if it's already in a bad state where it is getting bad close codes, that never happens, it just keeps getting bad codes, and keeps trying to reconnect, even though the last method the bot called was disconnect. This results in some other bad behavior as explained with the below situation:

First I waited until it was in the bad state described above, and got it into the state I just described where the last thing I called was disconnect. Then while it was in one of its long pauses after it says "Disconnected from voice channel 175384431383543809... Reconnecting in 1023.85s", I triggered a command (`?summon`) that calls connect().

<details>
  <summary>I got this log</summary>

```
INFO:discord.voice_client:Connecting to voice channel 175384431383543809...
INFO:discord.voice_client:Starting voice handshake...
INFO:discord.voice_client:_create_socket called for channel 175384431383543809
INFO:discord.voice_client:Voice handshake complete. Endpoint found us-west640.discord.media (IP: 162.159.138.234)
INFO:discord.gateway:selected the voice protocol for use (xsalsa20_poly1305_lite)
INFO:discord.gateway:received secret key for voice connection
INFO:discord.voice_client:end of connect() for channel 175384431383543809
INFO:discord.voice_client:Starting loop for channel 175384431383543809, (reconnect=True) (id=0.5789144089966105)
finished connect to: 175384431383543809
INFO:discord.voice_client:Entering poll_voice_ws for channel=175384431383543809 (id=0.5789144089966105) (loopid=0.9866699842304079)
```
</details>

Seems good right? The bot successfully connected, and I was even able to play some audio in the channel and the bot seemed to be working just fine. Then a couple minutes later, the bot suddenly disconnected from the channel, and reconnected. It didnt respond to any audio commands, and it thought that it wasnt connected even though it looked like it was. After a while it dropped out of the channel.

<details>
  <summary>This is whats happening in the logs</summary>

```
INFO:discord.voice_client:The voice handshake is being terminated for Channel ID 175384431383543809 (Guild ID 175384431383543808)
INFO:discord.voice_client:Doin a disconnectaroni for (channel=175384431383543809) (loopid=0.16362357981923203)
INFO:discord.voice_client:Calling disconnect for channel=175384431383543809 (id=0.6783516900838339)
INFO:discord.voice_client:The voice handshake is being terminated for Channel ID 175384431383543809 (Guild ID 175384431383543808)
INFO:discord.voice_client:The voice client has been removed for Channel ID 175384431383543809 (Guild ID 175384431383543808)
INFO:discord.voice_client:Disconnecting from voice normally, close code 4014. (channel=175384431383543809), (id=0.5789144089966105), (loopid=0.9866699842304079)
INFO:discord.voice_client:Calling disconnect for channel=175384431383543809 (id=0.5789144089966105)
INFO:discord.voice_client:The voice handshake is being terminated for Channel ID 175384431383543809 (Guild ID 175384431383543808)
INFO:discord.voice_client:The voice client has been removed for Channel ID 175384431383543809 (Guild ID 175384431383543808)
INFO:discord.voice_client:Exiting poll_voice_ws for channel=175384431383543809, (id=0.5789144089966105) (loopid=0.9866699842304079)
INFO:discord.voice_client:Connecting to voice channel 175384431383543809...
INFO:discord.voice_client:Starting voice handshake...
```
</details>

The first thing to notice about the above logs is that there are two different ids. These are the randomly generated ids that I added that are created when a voice_client is initialized. One is from the original voice_client, which has been infinitly looping in poll_voice_ws for a while (id=0.6783516900838339), and the other is from the new voice_client, which started up when I called the `?summon` command (id=0.5789144089966105). It appears that there are 2 voice_clients existing for the same server/channel. What seemed to happen as a result of that is that the old voice_client called disconnect, which successfully closed the new client (code 4014 is a anormal close code), and then tried (and failed) to reconnect by itself. This means that once it gets in this bad state, I don't think theres a way for it to get out.

# TLDR

Based on the above, it looks to me like there are 2 main issues that I've found:

1. voice_client.poll_voice_ws can get in a bad state if given non-normal close codes, and has no way to correctly recover. As I've show in the last set of logs, starting up a new voice_client is able to successfully connect, although it does get killed by the first one. That means that it's possible to change the implementation to fix the bad state, but I'm not sure how.
2. There can be 2 instances of voice_client for the same server/guild. This doesn't seems like it is intended. They interfere with each other and can cause problems. Does anyone here know why/how this could happen?

Thanks for taking the time to look at this. It has been a big problem with my bot for a while and it's made the voice commands super inconsistent. Since it is hard to reproduce (idk how to manually trigger a bad close code), its super hard to debug. Please let me know if you have any advice of things to try or of extra logging to add. Since it usually starts happening at least once after 4+ hours of the bot running, I can try changes out and see if we can get more information. Thank you!

### PR DESCRIPTION
**Pull request by mdiller**
### Summary

Fixes #3994
This fixes a scenario where a bad disconnect code on a websocket can cause us to end up with a zombie voice client that never goes away. After thorough investigation, it appears the problem was that we were at some point losing our reference to the voice client, as it was getting removed from the list of connected voice clients. this would be fine, except that the `poll_voice_ws` was still looping, often forever. I realized the only place that we remove the voice client from the list is when calling `terminate_handshake()` and passing in `remove=True`. We were only doing this in 2 places. One when we disconnect with a normal close code (which makes sense, thats the happy path of an actual closed connection), and one in `start_handshake()`. The problem was that when `poll_voice_ws` gets caught in a loop with a bad close code (see #3994), it would continuously fail the `start_handshake()` call before (sometimes) succeeding. Failing the call even once meant the voice client got removed from the list, which meant we had no way to actually close it, as the reference to it was essentially lost. This meant it kept coming back and interfering with things even if you created a new voice client and connected to the channel with it. Removing the argument as I did in this pull request solves the problem, and lets us close the client if its misbehaving.

Let me know if this is the wrong way to do this or if something else needs to be tweaked here.

### CODE REVIEW

**Comment by Rapptz** [comment]
Work on VoiceClient is now done on a separate branch called `voice-protocol` that is redesigning and reimplementing the protocol almost from scratch. I recommend checking if the issue is still present there and if possible doing the PR there instead since the old implementation is going to be superseded.

**Comment by mdiller** [comment]
ok will do, thanks for the info!

**Comment by HeartBurzum** [comment]
The new branch fixes all the issues I was having but I found a new bug. If a client has been forcefully disconnected from a voice channel the VoiceClient isn't cleaned up because the on_voice_status_update event removes the channel and the voice_connect method fails. I have bypassed this for now by making the event check only overwrite self.channel if the data passed in includes a channel id. I'm not sure if this would break anything else but it seems to pass all my checks so far.

**Comment by Rapptz** [comment]
@HeartBurzum That should be fixed now -- good catch.
